docs(reference): cargo-package reference docs (#992)
canonical/craft-parts: docs/common/craft-parts/reference/plugins/cargo-package_plugin.rst

.. _craft_parts_cargo-package_plugin:

Cargo package plugin
====================

The Cargo package plugin can be used for Rust projects that are dependencies of
other Rust packages. It is a companion plugin meant to be used with the
:ref:`Rust plugin <craft_parts_rust_plugin>`. Use of this plugin makes rust

> builds in **all** parts happen offline.

Reviewer: ```suggestion
builds in *all* parts happen offline.
```

Author: Thanks; I see this now under [Bold vs. Italics](https://docs.ubuntu.com/styleguide/en#interacting-with-ui-elements)

.. _craft_parts_cargo-package_plugin-keywords:

Keywords
--------

In addition to the common :ref:`plugin <part-properties-plugin>` and
:ref:`sources <part-properties-sources>` keywords, this plugin provides the
following plugin-specific keywords:

cargo-package-features
~~~~~~~~~~~~~~~~~~~~~~
**Type:** list of strings

Features used to build optional dependencies.
This is equivalent to the ``--features`` option in Cargo.

cargo-package-cargo-command
~~~~~~~~~~~~~~~~~~~~~~~~~~~
**Type:** string

**Example:** ``cargo: /usr/bin/cargo-1.82``

What command to use as the ``cargo`` executable. Can be used if a custom

> version of cargo is needed.

Reviewer: ```suggestion
version of Cargo is needed.
```

.. _craft_parts_cargo-package_plugin-environment_variables:

Environment variables
---------------------

CARGO_REGISTRY_DIRECTORY
~~~~~~~~~~~~~~~~~~~~~~~~

> The location where cargo will publish the crate. This does not need to be changed

Reviewer: ```suggestion
The location where Cargo will publish the crate. This doesn't need to be changed
```

in most cases.

.. _cargo-details-begin:

Dependencies
------------

> cargo must already be installed on the build system in order to use this plugin.

Reviewer: ```suggestion
Cargo must already be installed on the build system in order to use this plugin.
```

> .. _cargo-details-end:

Reviewer: ```suggestion

.. _cargo-details-end:
```

How it works
------------

During the build step, the plugin performs the following actions:

#. It sets up the system to use a local craft-parts directory registry
#. It packages the crate
#. It publishes the crate to a local directory registry for use by other parts

> When setting up the system, it makes the craft-parts directory registry, which is
> the destination of all parts using this plugin, the default registry. It also installs
> an ``apt`` registry to allow dependent parts to collect dependencies from
> ``librust-*-dev`` packages if they choose. In this case, the final part's crate

Reviewer: ```suggestion
When setting up the system, it creates a ``craft-parts`` directory registry, which is
the destination of all parts using this plugin. It also installs
an ``apt`` registry to allow dependent parts to collect dependencies from
``librust-*-dev`` packages if they choose. In this case, the final part's crate
```

Reviewer: It's much better now, thanks!

.. _craft_parts_cargo-package_plugin:

Cargo package plugin
====================

The Cargo package plugin can be used for Rust projects that are dependencies of
other Rust packages. It is a companion plugin meant to be used with the

> :ref:`Rust plugin <craft_parts_rust_plugin>`. Use of this plugin makes rust

Reviewer 2: ```suggestion
:ref:`Rust plugin <craft_parts_rust_plugin>`. Use of this plugin makes Rust
```

--------

In addition to the common :ref:`plugin <part-properties-plugin>` and
:ref:`sources <part-properties-sources>` keywords, this plugin provides the
following plugin-specific keywords:

cargo-package-features
~~~~~~~~~~~~~~~~~~~~~~
**Type:** list of strings

Features used to build optional dependencies.
This is equivalent to the ``--features`` option in Cargo.

cargo-package-cargo-command
~~~~~~~~~~~~~~~~~~~~~~~~~~~
**Type:** string

**Example:** ``cargo: /usr/bin/cargo-1.82``

What command to use as the ``cargo`` executable. Can be used if a custom
version of Cargo is needed.

.. _craft_parts_cargo-package_plugin-environment_variables:

Environment variables
---------------------

CARGO_REGISTRY_DIRECTORY
~~~~~~~~~~~~~~~~~~~~~~~~

The location where Cargo will publish the crate. This doesn't need to be changed
in most cases.

.. _cargo-details-begin:

Dependencies
------------

Cargo must already be installed on the build system in order to use this plugin.

.. _cargo-details-end:

How it works
------------

During the build step, the plugin performs the following actions:

#. It sets up the system to use a local craft-parts directory registry
#. It packages the crate

> #. It publishes the crate to a local ``craft-parts`` directory registry for use by

Reviewer 2: Why is "craft-parts" in code ticks in one spot, but not the other?

Author: ```suggestion
#. It sets up the system to use a local ``craft-parts`` directory registry
#. It packages the crate
#. It publishes the crate to a local ``craft-parts`` directory registry for use by
```
